File: rename.js
const readline = require('readline');
const fs = require('fs');
const path = require('path');


const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});
rl.question('请输入文件夹： ', (answer) => {
    getDir(answer);
});



function getDir(catalogue) {
    const currentDir = process.cwd();
    const targetDir = path.join(currentDir, catalogue);

    fs.readdir(targetDir, (err, files) => {
        if (err) {
            console.error(`无法读取目录 ${catalogue}：`, err);
            console.log('准备退出...');
            process.exit();
            return;
        }
        // console.log(`目录 ${targetDir} 中的文件列表：`);
        // rl.question(`是否删除？（y/n） `, (answer) => {
        // if (answer.toLowerCase() === 'y') {
        // } else {
        //     console.log(`文件未被删除。`);
        //     rl.close();
        // }
        // });
        files.forEach(file => {
            // console.log(file);
            if (file.endsWith('.tc.ass') || file.endsWith('.TC.ass')) {
                const filePath = path.join(targetDir, file);
                fs.unlink(filePath, err => {
                    if (err) {
                        console.error(`无法删除文件 ${filePath}：`, err);
                    } else {
                        console.log(`成功删除文件 ${filePath}`);
                    }
                });
            } else {
                console.log(file);
            }
        });
        reName(catalogue)
    });
}

function reName(catalogue) {
    const currentDir = process.cwd();
    const targetDir = path.join(currentDir, catalogue);
    rl.question('请输入新名称： ', (answer) => {
        console.log(`你输入的名称是： ${answer}`);

        fs.readdir(targetDir, (err, files) => {
            // 对文件按名称进行排序
            files.sort();
            files.forEach((file, index) => {
                // 重命名文件
                const oldPath = path.join(targetDir, file);
                const replaceEpisodeNumber = (fileName, newEpisodeNumber) => {
                    const regexArray = [/- (\d+)/, /\[(\d+)\]/, /第(\d+)/];
                    for (const regex of regexArray) {
                        const match = fileName.match(regex);
                        if (match) {
                            const episodeNumber = parseInt(newEpisodeNumber);
                            const formattedEpisodeNumber = episodeNumber < 10 ? `0${episodeNumber}` : `${episodeNumber}`;
                            // 只替换匹配项里的集数
                            return fileName.replace(match[1], formattedEpisodeNumber);
                        }
                    }
                    return '未找到匹配的剧集编号';
                };

                const newFileName = replaceEpisodeNumber(answer, index + 1);
                if (newFileName == '未找到匹配的剧集编号') {
                    console.log(newFileName, 'newFileName');
                } else {
                    const newPath = path.join(targetDir, newFileName + '.ass');
                    fs.rename(oldPath, newPath, (err) => {
                        if (err) {
                            console.error(`无法重命名文件 ${file}：`, err);
                        } else {
                            console.log(`成功重命名文件 ${file} 为 ${newFileName}`);
                        }
                    });
                }

            });
        });
        rl.close();
    });


}

// 已覆盖例子
// [Sakurato.Sub&FS-Raw] Tate no Yuusha no Nariagari - 01 (HEVC-10Bit-2160P AAC)
// [VCB-Studio] Yamada-kun to 7-nin no Majo [01][Ma10p_1080p][x265_flac]
// [Snow-Raws] DARKER THAN BLACK 黒の契約者 第01話 (BD 1920x1080 HEVC-YUV420P10 FLACx2)

File: rename.py
import os
import re
import tkinter as tk
from tkinter import filedialog


def get_dir():
    root = tk.Tk()
    root.withdraw()

    # 弹出对话框选择文件夹，默认为当前目录
    catalogue = filedialog.askdirectory(initialdir=os.getcwd())

    if not catalogue:
        print("未选择文件夹，准备退出...")
        return

    try:
        files = os.listdir(catalogue)
        for file in files:
            if file.endswith('.tc.ass') or file.endswith('.TC.ass'):
                file_path = os.path.join(catalogue, file)
                os.remove(file_path)
                print(f"成功删除文件 {file_path}")
            else:
                print(file)

        re_name(catalogue)
    except OSError as e:
        print(f"无法读取目录 {catalogue}：", e)
        print('准备退出...')


def re_name(catalogue):
    root = tk.Tk()
    root.withdraw()

    # 弹出对话框输入新名称
    new_name = tk.simpledialog.askstring("输入新名称", "请输入新名称：")

    if not new_name:
        print("未输入新名称，准备退出...")
        return

    print(f"你输入的名称是： {new_name}")

    try:
        files = os.listdir(catalogue)
        files.sort()
        for index, file in enumerate(files):
            old_path = os.path.join(catalogue, file)

            def replace_episode_number(file_name, new_episode_number):
                regex_array = [r'- (\d+)', r'\[(\d+)\]', r'第(\d+)']
                for regex in regex_array:
                    match = re.search(regex, file_name)
                    if match:
                        episode_number = int(new_episode_number)
                        formatted_episode_number = f"{episode_number:02d}" if episode_number < 10 else f"{episode_number}"
                        return re.sub(match.group(1), formatted_episode_number, file_name)
                return '未找到匹配的剧集编号'

            new_file_name = replace_episode_number(new_name, index + 1)
            if new_file_name == '未找到匹配的剧集编号':
                print(new_file_name, 'new_file_name')
            else:
                new_path = os.path.join(catalogue, new_file_name + '.ass')
                os.rename(old_path, new_path)
                print(f"成功重命名文件 {file} 为 {new_file_name}")

    except OSError as e:
        print(f"无法重命名文件 {file}：", e)

get_dir()

# 已覆盖例子
# [Sakurato.Sub&FS-Raw] Tate no Yuusha no Nariagari - 01 (HEVC-10Bit-2160P AAC)
# [VCB-Studio] Yamada-kun to 7-nin no Majo [01][Ma10p_1080p][x265_flac]
# [Snow-Raws] DARKER THAN BLACK 黒の契約者 第01話 (BD 1920x1080 HEVC-YUV420P10 FLACx2)